src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -446,6 +448,17 @@ private static unsafe long GetRandomNumber()
             return *((long*)&g);
         }
 
+        private static bool ValidateSetCurrent(Activity activity)
+        {
+            bool canSet = activity == null || (activity.Id != null && !activity.isFinished);
```

Comment:
There are potential races here, but I believe the right answer is to simply tolerate them.

Reply:
Agree, concurrent usage and modification of `Activity.Current` should be extremely rare and does not look like a valid scenario